README.md
<!--
SPDX-FileCopyrightText: 2021 Ian2020, et. al. <https://github.com/Ian2020>

SPDX-License-Identifier: CC-BY-SA-4.0

Keep your accounts offline

For full copyright information see the AUTHORS file at the top-level
directory of this distribution or at
[AUTHORS](https://github.com/Ian2020/offlinebooks/AUTHORS.md)

This work is licensed under the Creative Commons Attribution 4.0 International
License. You should have received a copy of the license along with this work.
If not, visit http://creativecommons.org/licenses/by/4.0/ or send a letter to
Creative Commons, PO Box 1866, Mountain View, CA 94042, USA.
-->

# offlinebooks

Keep your accounts offline.

Offlinebooks is a cmdline tool that downloads your Xero financial accounts for
local backup. It is beta quality software and only Linux (GNOME desktop) is
supported at present. Use with caution. The code repository is
at [offlinebooks](https://github.com/Ian2020/offlinebooks).

We are not associated with the Xero organisation and this software is
unsupported. See [Licensing](#licensing) for more details.

## Table of Contents

* [Install](#install)
* [Background](#background)
* [Usage](#usage)
* [Known Issues](#known-issues)
* [Roadmap](#roadmap)
* [Contributing](#contributing)
* [Licensing](#licensing)
* [Attribution](#attribution)

## Install

Prerequisites:

* Linux (GNOME desktop)
* Python 3
* [Xoauth](https://github.com/XeroAPI/xoauth)

Install the latest version via pip:

```bash
pip install offlinebooks
```

This will install a cmdline tool `offlinebooks`. For usage see below.

## Background

I just wanted a simple way to pull my data from the Xero API and save it locally
for backup.

We save each entity as JSON in its own file, named by its unique ID (where
available) in a simple dir tree structure. This allows for easy processing with
other tools and is also suitable for source control. If an entity has attachments
those are saved next to the entity under a dir named `[entity ID]_attachments`

Data is saved separately for each tenant (organisation) under their tenant id
and under their human-readable tenant name as a symlink for convenience:

```text
$XDG_DATA_HOME/offlinebooks/
├── tenantName
│   ├── Demo Company (UK) -> $XDG_DATA_HOME/offlinebooks/tenantId/b3b892ur-02i8-4842-8bx2-85696h032kz2
│   └── ...
└── tenantId
    ├── b3b892ur-02i8-4842-8bx2-85696h032kz2
    │   ├── accounts
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7.json
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7_attachments
    │   │   │   └── ...
    │   │   └── ...
    │   ├── brandingthemes
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7.json
    │   │   └── ...
    │   ├── contacts
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7.json
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7_attachments
    │   │   │   └── ...
    │   │   └── ...
    │   ├── currencies
    │   │   ├── GBP.json
    │   │   └── ...
    │   ├── invoices
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7.json
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7_attachments
    │   │   │   └── ...
    │   │   └── ...
    │   ├── items
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7.json
    │   │   └── ...
    │   ├── journals
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7.json
    │   │   └── ...
    │   ├── organisations
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7.json
    │   │   └── ...
    │   ├── taxrates
    │   │   ├── 15% (VAT on Capital Purchases).json
    │   │   └── ...
    │   └── users
    │   │   ├── a6r01id3-690x-7edt-8pd2-c873245y38v7.json
    │   │   └── ...
    └── ...
```

Data covered so far:

* Accounts
* Bank Transactions
* Bank Transfers
* Branding theme
* Contacts
* Currencies
* Invoices
* Items
* Journals
* Manual Journals
* Organisations, EXCEPT:
  * Organisation actions
  * CIS Settings (UK)
* Payments
* Users

## Usage

The first time around you will need to add offlinebooks as a PKCE app to your Xero
account and grant it permissions:

* Thanks to
  [JWealthall](https://github.com/JWealthall) you can simply follow his excellent
  [PKCE How To for Xero OAuth2
  API](https://github.com/JWealthall/XeroOAuth2ApiPkceHowTo) with the following
  changes:
  * Use `offlinebooks` as both the 'App name' in Xero and the clientname when
    running `xoauth setup`.
  * For company or application URL you can put whatever you like or can I suggest
    the [offlinebooks project page on PyPI](https://pypi.org/project/offlinebooks/)
  * When it gets to adding scopes we need `openid` to authorise the app and
    `offline_access` to refresh our token on expiry. Beyond that
    we just need read permissions for each entity we download. DO NOT ADD ANY
    WRITE PERMISSIONS, they are not needed. The following should suffice:

```bash
xoauth setup add-scope offlinebooks \
  openid \
  offline_access \
  accounting.transactions.read \
  accounting.contacts.read \
  accounting.journals.read \
  accounting.settings.read \
  files.read \
  accounting.reports.read \
  accounting.attachments.read
```

* Once you've completed the How To then you are ready to run offlinebooks:

```bash
offlinebooks
```

If it completes without error you'll find a new dir at
`$XDG_DATA_HOME/offlinebooks` (probably `~/.local/share/offlinebooks`)
containing the downloaded data for you to explore (see above).

If it hits the Xero API minute rate-limit it'll pause for a time interval (as
supplied by the API) before continuing the run. You'll see some output to
indicate this, for example:

```text
Xero API rate-limit exceeded for calls per minute, will pause for 43.0s
API reported the following remaining limits:
  app minute : 9911
  daily      : 2718
```

## Troubleshooting

* `No such file or directory: '~/.xoauth/xoauth.json'` - have you run
  xoauth as detailed above?

## Known Issues

* Possibly fixed since v0.0.8: still get API rate-limits exceeded sometimes:

```bash
Xero API rate-limit exceeded for calls per minute, will pause for 14.0s
API reported the following remaining limits:
  app minute : 9984
  daily      : 4794
Traceback (most recent call last):
  File "/home/user/.local/bin/offlinebooks", line 8, in <module>
    sys.exit(main())
  File "/home/user/.local/lib/python3.9/site-packages/offlinebooks/main.py", line 329, in main
    for entity in fetcher.fetch():
  File "/home/user/.local/lib/python3.9/site-packages/offlinebooks/main.py", line 237, in all_generator
    items = func.all()
  File "/home/user/.local/lib/python3.9/site-packages/xero/basemanager.py", line 264, in wrapper
    raise XeroRateLimitExceeded(response, payload)
xero.exceptions.XeroRateLimitExceeded: please wait before retrying the xero api,the limit exceeded is: minute
```

  This is hard to reproduce. I have added (24th Mar '22) an extra second to our
  wait time in case of rounding issues though I have seen this exception occur
  before we decide to wait. Also added extra logging in case we are reaching
  our retry stop limit without realising.

* Intermittent exception 'oauthlib.oauth2.rfc6749.errors.InvalidGrantError' -
  not sure what causes this. If you do a `xoauth connect` then run again it
  should work.
* A long delay then a failure with exception
  `xero.exceptions.XeroExceptionUnknown: None` whilst calling func.filter with
  paging. If you `xoauth connect` then run again it works.
* We depend on `secret-tool` to retrieve the token which `xoauth` has saved,
  this is GNOME-specific.
* Attachments are only re-fetched if they differ in size to avoid large
  downloads. There's a small chance that an updated attachment which is the same
  size as the original is not fetched.
* We do not yet download anything outside of the Accounting API, that leaves
  these APIs untouched:
  * Payroll
  * Assets
  * Files
  * Projects
  * Bank Feeds
  * Xero HQ
  * Practice Manager
  * WorkflowMax
* Within the Accounting API we do not yet save the following (for some we also
  show in parenthesis where these are found in the Xero web GUI):
  * Batch Payments
  * Budgets
  * Contact Groups
  * Credit Notes
  * Employees
  * History and Notes
  * Invoice Reminders
  * Linked Transactions
  * Payment Services
  * Prepayments
  * Purchase Orders
  * Quotes
  * Repeating Invoices (Business-Invoices-Repeating)
  * Reports (Accounting-Reports)
  * Tracking Categories (Accounting-Advanced-Tracking Categories)

## Roadmap

In vague priority order:

* Download remaining data not yet supported from Accounting API above.
  * How are VAT returns covered by the API?
* Allow user to limit tenant(s). Introduce config file in XDG friendly location
* Allow user to specify repo path also
* We should report at the end on API usage if requested: 'Each API response you
  receive will include the X-DayLimit-Remaining, X-MinLimit-Remaining and
  X-AppMinLimit-Remaining headers telling you the number of remaining against
  each limit.' This can be got from the final response we receive.
* Fetch in parallel
* We assume journals start at 1, i.e. setting offset=0 which means querying
  JournalNumber>0. Is this definite?
* Our use of tenacity and handling API limits:
  * We repeat the retry decorator on every function containing an API call.
    Better to remove this duplication.
  * It would be better not to reach into the response and get headers,
    perhaps pyxero's XeroRateLimitExceeded exception could hold the data
    instead.

## Contributing

It's great that you're interested in contributing. Please ask questions by
raising an issue, be sure to get in touch before raising PRs to avoid wasted
work. For full details see [CONTRIBUTING.md](CONTRIBUTING.md)

## Licensing

We declare our licensing by following the REUSE specification - copies of
applicable licenses are stored in the LICENSES directory. Here is a summary:

* All source code is licensed under AGPL-3.0-or-later.
* Anything else that is not executable, including the text when extracted from
  code, is licensed under CC-BY-SA-4.0.

For more accurate information, check individual files.

Offlinebooks is free software: you can redistribute it and/or modify it under the
terms of the GNU Affero General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.

## Attribution

"Xero" is a trademark of Xero Limited.


CHANGELOG.md
<!--
SPDX-FileCopyrightText: 2021 Ian2020, et. al. <https://github.com/Ian2020>

SPDX-License-Identifier: CC-BY-SA-4.0

Keep your accounts offline

For full copyright information see the AUTHORS file at the top-level
directory of this distribution or at
[AUTHORS](https://github.com/Ian2020/offlinebooks/AUTHORS.md)

This work is licensed under the Creative Commons Attribution 4.0 International
License. You should have received a copy of the license along with this work.
If not, visit http://creativecommons.org/licenses/by/4.0/ or send a letter to
Creative Commons, PO Box 1866, Mountain View, CA 94042, USA.
-->

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.8](https://github.com/ian2020/offlinebooks/releases/tag/v0.0.8) - 2022-07-06

### Fixed

* Possible fix for intermittent XeroRateLimitExceeded exceptions.
  * Tenacity does not work with generators, we have to apply it to functions
    that call the generators:
    ["Retry" from tenacity module doesn't work with a generator](https://stackoverflow.com/a/61571713/28170)
  * Tenacity can say in its `retry_state` that a method threw an exception but
    still give it an outcome of 'finished'. We were filtering on 'failed' only
    which meant we were ignoring some API exceptions and letting them bubble
    out. Now we just focus on whether a method had an exception regardless of
    the outcome state.

### Added

* Set changelog and bug tracker URLs in PyPI.

## [0.0.7](https://github.com/ian2020/offlinebooks/releases/tag/v0.0.7) - 2022-03-24

### Added

To try and track down what's causing the occasional XeroRateLimitExceeded
exception we sometimes see the following measures were added:

* Extra logging if we reach the limit of our retries against the API.
* Add a one second buffer to the wait time we're given by the API.

## [0.0.6](https://github.com/ian2020/offlinebooks/releases/tag/v0.0.6) - 2021-12-14

### Fixed

* Now we can cope if Xero API responds that we have breached the calls per minute
  limit. We back off for a time interval as supplied by them and then retry. We
  output this event to the console so users can see the delay.

## [0.0.5](https://github.com/ian2020/offlinebooks/releases/tag/v0.0.5) - 2021-11-10

### Added

* Update dependencies - latest pyxero 0.9.3 fixes some important bugs in
  particular:
  [Release v0.9.3 · freakboy3742/pyxero](https://github.com/freakboy3742/pyxero/releases/tag/v0.9.3)
* Instructions on adding new entities and other minor documentation.

## [0.0.4](https://github.com/ian2020/offlinebooks/releases/tag/v0.0.4) - 2021-07-09

### Added

* Bank transactions, transfers, manual journals and payments now all saved.

### Changed

* Now classified under accounting topic for PyPI.

## [0.0.3](https://github.com/ian2020/offlinebooks/releases/tag/v0.0.3) - 2021-05-27

### Changed

* Journal lines in journals are now sorted by ID before we save them. The API
  brings them back in varying order and this causes nuisance changes in source
  control otherwise.

## [0.0.2](https://github.com/Ian2020/offlinebooks/releases/tag/v0.0.2) - 2021-05-26

### Added

* Attachments are now saved for invoices, contacts and accounts. These will
  appear in a dir `[entity ID]_attachments` alongside the entity they pertain to.

### Changed

* Data is now stored in `~/.local/share/offlinebooks` instead of
  `~/.offlinebooks`. This follows the [XDG Directory
  Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)


offlinebooks/__init__.py
# SPDX-FileCopyrightText: 2021 Ian2020, et. al. <https://github.com/Ian2020>
#
# SPDX-License-Identifier: AGPL-3.0-or-later
#
# Keep your accounts offline
#
# For full copyright information see the AUTHORS file at the top-level
# directory of this distribution or at
# [AUTHORS](https://github.com/Ian2020/offlinebooks/AUTHORS.md)
#
# This program is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

offlinebooks/main.py
#! /bin/env python

# SPDX-FileCopyrightText: 2021 Ian2020, et. al. <https://github.com/Ian2020>
#
# SPDX-License-Identifier: AGPL-3.0-or-later
#
# Keep your accounts offline
#
# For full copyright information see the AUTHORS file at the top-level
# directory of this distribution or at
# [AUTHORS](https://github.com/Ian2020/offlinebooks/AUTHORS.md)
#
# This program is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

import json
import logging
import os
import subprocess
from collections import namedtuple
from pathvalidate import sanitize_filename
from tenacity import (retry, stop_after_delay, stop_after_attempt)
from xero.auth import OAuth2Credentials
from xero import Xero
from xero.exceptions import XeroRateLimitExceeded


Entity = namedtuple("Entity", "id data")

# #########################
# TENACITY CUSTOM CALLBACKS
# #########################


# [Re-raising RetryError from retry_error_callback · Issue #296 · jd/tenacity]
# (https://github.com/jd/tenacity/issues/296)
def retry_on_error(retry_state):
    logger = logging.getLogger(__name__)
    logger.exception(
        "Max retries exceeded",
        extra={"stats": retry_state.retry_object.statistics},
    )
    raise retry_state.retry_object.retry_error_cls(retry_state.outcome) \
          from retry_state.outcome.exception()


def retry_if_minute_rate_limit_exceeded(retry_state):
    # This gets called on every method return and/or exception
    return \
        (isinstance(retry_state.outcome.exception(), XeroRateLimitExceeded)
         and
         retry_state.outcome.exception().response.headers.get(
             "X-Rate-Limit-Problem") == 'minute')


def wait_till_api_says_retry(retry_state) -> float:
    default = 60.0
    if isinstance(retry_state.outcome.exception(), XeroRateLimitExceeded):
        default = \
                float(
                    retry_state.outcome.exception().response.headers.get(
                        "Retry-After"))
    # Add a tiny buffer just in case
    return default + 1


def before_sleep_log(retry_state):
    logger = logging.getLogger(__name__)
    logger.warning(f"Xero API rate-limit exceeded for calls per minute, "
                   f"will pause for {retry_state.next_action.sleep}s")
    if isinstance(retry_state.outcome.exception(), XeroRateLimitExceeded):
        headers = retry_state.outcome.exception().response.headers
        logger.warning("API reported the following remaining limits:")
        logger.warning(
            f"  app minute : {headers.get('X-AppMinLimit-Remaining')}")
        logger.warning(
            f"  daily      : {headers.get('X-DayLimit-Remaining')}")


# ########
# FETCHERS
# ########


class BankTransactions:
    def __init__(self, xero):
        self.name = "banktransactions"
        self.fetch = lambda: paged_generator(xero.banktransactions,
                                             "BankTransactionID")


class BankTransfers:
    def __init__(self, xero):
        self.name = "banktransfers"
        self.fetch = lambda: all_generator(xero.banktransfers,
                                           "BankTransferID")


class ManualJournals:
    def __init__(self, xero):
        self.name = "manualjournals"
        self.fetch = lambda: paged_generator(xero.manualjournals,
                                             "ManualJournalID")


class Payments:
    def __init__(self, xero):
        self.name = "payments"
        self.fetch = lambda: paged_generator(xero.payments,
                                             "PaymentID")


class Contacts:
    def __init__(self, xero):
        self.name = "contacts"
        self.fetch = lambda: paged_generator(xero.contacts, "ContactID",
                                             includeArchived=True)


class Journals:
    def __init__(self, xero):
        self.name = "journals"
        self.xero = xero

    def fetch(self):
        more = True
        offset = 0
        while more:
            items = self.xero.journals.filter(offset=offset)
            if len(items) == 0:
                more = False
            else:
                for item in items:
                    # Sort journal lines as their order is not guaranteed by
                    # the API
                    item['JournalLines'] = sorted(
                        item['JournalLines'], key=lambda k: k['JournalLineID'])
                    yield Entity(item['JournalID'], item)
                offset = items[-1]['JournalNumber']


class Invoices:
    def __init__(self, xero):
        self.name = "invoices"
        self.fetch = lambda: paged_generator(xero.invoices, "InvoiceID")


class Accounts:
    def __init__(self, xero):
        self.name = "accounts"
        self.fetch = lambda: all_generator(xero.accounts, "AccountID")


class Currencies:
    def __init__(self, xero):
        self.name = "currencies"
        self.fetch = lambda: all_generator(xero.currencies, "Code")


class Items:
    def __init__(self, xero):
        self.name = "items"
        self.fetch = lambda: all_generator(xero.items, "ItemID")


class Organisations:
    def __init__(self, xero):
        self.name = "organisations"
        self.fetch = lambda: all_generator(xero.organisations,
                                           "OrganisationID")


class TaxRates:
    def __init__(self, xero):
        self.name = "taxrates"
        self.fetch = lambda: all_generator(xero.taxrates, "Name")


class Users:
    def __init__(self, xero):
        self.name = "users"
        self.fetch = lambda: all_generator(xero.users, "UserID")


class BrandingThemes:
    def __init__(self, xero):
        self.name = "brandingthemes"
        self.fetch = lambda: all_generator(xero.brandingthemes,
                                           "BrandingThemeID")


def get_token():
    return eval(subprocess.run(
        ["/usr/bin/secret-tool",
            "lookup",
            "service",
            "com.xero.xoauth",
            "username",
            "offlinebooks:token_set"],
        capture_output=True,
        text=True
        ).stdout.strip())


def get_client_id():
    with open(os.path.join(os.path.expanduser("~"),
                           ".xoauth",
                           "xoauth.json"), "r") as file:
        xoauth = eval(file.read())
    return xoauth['offlinebooks']['ClientId']


def paged_generator(func, id_field, **kwargs):
    more = True
    page_no = 1
    while more:
        items = func.filter(page=page_no, **kwargs)
        if len(items) == 0:
            more = False
        else:
            for item in items:
                yield Entity(item[id_field], item)
            page_no += 1


def all_generator(func, id_field):
    items = func.all()
    for item in items:
        yield Entity(item[id_field], item)


def get_repo():
    # https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
    if "XDG_DATA_HOME" in os.environ:
        data_home = os.environ["XDG_DATA_HOME"]
    else:
        data_home = os.path.join(os.path.expanduser("~"), ".local", "share")
    return os.path.join(data_home, "offlinebooks")


@retry(retry=retry_if_minute_rate_limit_exceeded,
       retry_error_callback=retry_on_error,
       stop=stop_after_delay(90) | stop_after_attempt(5),
       before_sleep=before_sleep_log,
       wait=wait_till_api_says_retry)
def process_attachments(xero, entity, entity_attachments_dir):
    attachmentContainer = xero.invoices.get_attachments(entity.id)
    for attach in attachmentContainer['Attachments']:
        download = False
        attach_filename = os.path.join(entity_attachments_dir,
                                       attach['FileName'])
        if os.path.exists(attach_filename):
            # TODO: This check may not be sufficient if an attachment is
            # updated but the same size as before. We should also check if its
            # unique ID has changed.
            if os.path.getsize(attach_filename) != attach['ContentLength']:
                download = True
        else:
            download = True
        if download:
            os.makedirs(entity_attachments_dir, exist_ok=True)
            with open(attach_filename, "wb") as f:
                # TODO: This does not seem robust, to fetch by filename
                xero.invoices.get_attachment(entity.id, attach['FileName'], f)


@retry(retry=retry_if_minute_rate_limit_exceeded,
       retry_error_callback=retry_on_error,
       stop=stop_after_delay(90) | stop_after_attempt(5),
       before_sleep=before_sleep_log,
       wait=wait_till_api_says_retry)
def retry_generator(func):
    # Tenacity cannot work with generators so we must apply it to functions
    # that call the generators
    # https://stackoverflow.com/a/61571713/28170
    results = list(func())
    return results


def main():
    client_id = get_client_id()
    token = get_token()

    # Even tho PKCE we set client_secret to avoid bug of exception on refresh
    credentials = OAuth2Credentials(client_id,
                                    token=token,
                                    client_secret="RUBBISH")

    # Don't care if more scopes granted than requested
    # https://stackoverflow.com/a/51643134
    os.environ["OAUTHLIB_RELAX_TOKEN_SCOPE"] = "1"
    if credentials.expired():
        credentials.refresh()

    tenants = credentials.get_tenants()
    xero = Xero(credentials)
    repo = get_repo()

    for tenant in tenants:
        credentials.tenant_id = tenant['tenantId']
        repo_tenantId = os.path.join(repo,
                                     'tenantId',
                                     sanitize_filename(tenant['tenantId']))
        repo_tenantName = os.path.join(repo, 'tenantName')
        os.makedirs(repo_tenantId, exist_ok=True)
        os.makedirs(repo_tenantName, exist_ok=True)

        tenant_sym = os.path.join(repo_tenantName,
                                  sanitize_filename(tenant['tenantName']))
        if not os.path.isdir(tenant_sym):
            os.symlink(repo_tenantId, tenant_sym)

        fetchers = [BankTransactions(xero),
                    BankTransfers(xero),
                    ManualJournals(xero),
                    Payments(xero),
                    Contacts(xero),
                    Journals(xero),
                    Invoices(xero),
                    Accounts(xero),
                    Currencies(xero),
                    Items(xero),
                    Organisations(xero),
                    TaxRates(xero),
                    Users(xero),
                    BrandingThemes(xero),
                    ]

        for fetcher in fetchers:
            save_dir = os.path.join(repo_tenantId, fetcher.name)
            os.makedirs(save_dir, exist_ok=True)

            for entity in retry_generator(fetcher.fetch):
                safe_entity_id = sanitize_filename(entity.id)
                entity_path = os.path.join(
                                    save_dir, f"{safe_entity_id}.json")
                entity_attachments_dir = os.path.join(
                                    save_dir, f"{safe_entity_id}_attachments")
                with open(entity_path, "w") as file:
                    json.dump(entity.data,
                              file,
                              indent=4,
                              sort_keys=True,
                              default=str)
                if 'HasAttachments' in entity.data and \
                   bool(entity.data['HasAttachments']):
                    process_attachments(xero, entity, entity_attachments_dir)


if __name__ == "__main__":
    main()
